src/Futhark/CodeGen/Backends/CWebGPU.hs

```diff
@@ -0,0 +1,382 @@
+    freeCall = asyncCall
+      ("futhark_free_" <> name) True ["this.ctx", "arr"]
+    valuesCall = asyncCall
+      ("futhark_values_" <> name) True ["this.ctx", "arr", "data"]
+    valuesArrayCopy = primWasmArrayCopy typ sign "idx" "len"
+    exports = ["futhark_new_" <> name, "futhark_free_" <> name,
+               "futhark_values_" <> name, "futhark_shape_" <> name]
+
+primWasmHeap :: PrimType -> Signedness -> T.Text
+primWasmHeap Bool _ = "HEAP8"
+primWasmHeap (IntType Int8) Unsigned = "HEAPU8"
+primWasmHeap (IntType Int8) Signed = "HEAP8"
+primWasmHeap (IntType Int16) Unsigned = "HEAPU16"
+primWasmHeap (IntType Int16) Signed = "HEAP16"
+primWasmHeap (IntType Int32) Unsigned = "HEAPU32"
+primWasmHeap (IntType Int32) Signed = "HEAP32"
+primWasmHeap (IntType Int64) Unsigned = "HEAPU64"
+primWasmHeap (IntType Int64) Signed = "HEAP64"
+primWasmHeap (FloatType Float16) _ = error "f16 wasm heap op: unimplemented"
+primWasmHeap (FloatType Float32) _ = "HEAPF32"
+primWasmHeap (FloatType Float64) _ = "HEAPF64"
+primWasmHeap Unit _ = error "unit wasm heap op: unimplemented"
+
+primWasmArrayCopy :: PrimType -> Signedness -> T.Text -> T.Text -> T.Text
+primWasmArrayCopy t@(IntType _) sign idx len =
+  let heapVar = primWasmHeap t sign
+   in [text|Array.from(this.m.${heapVar}.subarray(${idx}, ${idx} + ${len}))|]
+primWasmArrayCopy t@(FloatType _) sign idx len =
+  let heapVar = primWasmHeap t sign
+   in [text|Array.from(this.m.${heapVar}.subarray(${idx}, ${idx} + ${len}))|]
+primWasmArrayCopy Bool _ idx len =
+  [text|Array.from(this.m.HEAP8.subarray(${idx}, ${idx} + ${len})) .map((x) => x != 0)|]
+primWasmArrayCopy Unit _ _ _ = error "unit copy from WASM heap: unimplemented"
+
+-- | Compile the program to C with calls to WebGPU, along with a JS wrapper
+-- library.
+compileProg :: (MonadFreshNames m) => T.Text -> Prog GPUMem
+            -> m (ImpGen.Warnings, (GC.CParts, T.Text, [T.Text]))
+compileProg version prog = do
+  ( ws,
+    Program wgsl_code wgsl_prelude macros kernels params failures prog'
+    ) <- ImpGen.compileProg prog
+  c <- GC.compileProg
+        "webgpu"
+        version
+        params
+        operations
+        (mkBoilerplate (wgsl_prelude <> wgsl_code) macros kernels [] failures)
+        webgpu_includes
+        (Space "device", [Space "device", DefaultSpace])
+        cliOptions
+        prog'
+  let (js, exports) = jsBoilerplate prog' (GC.cJsonManifest c)
+  pure (ws, (c, js, exports))
+  where
+    operations :: GC.Operations HostOp ()
+    operations =
+      gpuOperations
+        { GC.opsMemoryType = webgpuMemoryType }
+    webgpu_includes =
+      [untrimming|
+       #ifdef USE_DAWN
+         #include <dawn/webgpu.h>
+       #else
+         #include <webgpu/webgpu.h>
+         #include <emscripten.h>
+         #include <emscripten/html5_webgpu.h>
+       #endif
+      |]
+
+-- | As server script. Speaks custom server protocol to local Python server
```

Comment:
I suggest not using the term "server protocol" to reference anything but the "Futhark server protocol" (which should be understood as "(Futhark server) protocol". The term "server protocol" on its own is either wrong or redundant. Just call a  "custom protocol" or "browser test protocol" if it must have a name.